Add dependency checksums to quarkus caching extension (#1076)
gradle/develocity-build-config-samples: quarkus-build-caching-extension/README.md

# Custom Maven Extension to make Quarkus build goal cacheable
This Maven extension allows to make the [Quarkus Maven plugin](https://quarkus.io/guides/quarkus-maven-plugin) `build` goal cacheable.

This project performs programmatic configuration of the [Develocity Build Cache](https://docs.gradle.com/enterprise/maven-extension/#using_the_build_cache) through a Maven extension. See [here](https://docs.gradle.com/enterprise/maven-extension/#custom_extension) for more details.
This project is derived from [the CCUD Maven extension](https://github.com/gradle/common-custom-user-data-maven-extension).

*Note:*<br>

> A native executable can be a very large file, and copying it from/to the local cache, or transferring it from/to the remote cache can be an expensive operation that has to be balanced with the caching benefit

Reviewer: ```suggestion
A native executable can be a very large file. Copying it from/to the local cache, or transferring it from/to the remote cache can be an expensive operation that has to be balanced with the duration of the work being avoided.
```

## Requirements
- Quarkus 3.2.4 and above which brings [track-config-changes goal](https://quarkus.io/guides/config-reference#tracking-build-time-configuration-changes-between-builds)

## Limitations
- Only the `native`, `uber-jar`, `jar` and `legacy-jar` [packaging types](https://quarkus.io/guides/maven-tooling#quarkus-package-pkg-package-config_quarkus.package.type) can be made cacheable
- The `native` packaging is cacheable only if the in-container build strategy (`quarkus.native.container-build=true`) is configured along with a fixed build image (`quarkus.native.builder-image`).

**Note:**<br>

> - When the in-container build strategy is used as a fallback due to GraalVM requirements not met for instance, the caching feature will still be disabled. The recommendation is to explicitly set the in-container strategy (`quarkus.native.container-build=true`) to benefit from caching

Reviewer: ```suggestion
- When the in-container build strategy is used as a fallback the caching feature will be disabled. The fallback may happen due to GraalVM requirements not met. The recommendation is to explicitly set the in-container strategy (`quarkus.native.container-build=true`) to benefit from caching
```

</plugin>
```

## Improving the cache hit rate
The file `.quarkus/quarkus-prod-config-dump` generated by the Quarkus build `goal` must be present to allow a cache hit.

In order to increase the cache hit rate, it is possible to check-in the file in the source code repository.

> in continuous integration environment, an alternative is to have a cache restoring the file.

Reviewer: `an alternative is to have a cache restoring the file.`

The word cache here can be misleading. I assume you mean some sort of CI store/restore functionality? Not the build cache?

Author: Correct, I'll update with
> in continuous integration environment, an alternative is to have the file restored (see [this option](https://github.com/marketplace/actions/cache#restoring-and-saving-cache-using-a-single-action) for GitHub actions as an example).

Reviewer: i like this

- as a maven property:
`<gradle.quarkus.extension.config.file>.quarkus/extension-local.properties</gradle.quarkus.extension.config.file>`

## Implementation details

> The instructions required to make the Quarkus `build` goal cacheable are isolated to the [QuarkusBuildCache](./src/main/java/com/gradle/QuarkusBuildCache.java) class.

Reviewer: by instructions here do you mean instructions for the consumer of the caching utility? Or do you mean the code to make the Quarkus build goal cacheable?

Author: I meant the code, let me rephrase to clarify with
> The logic to make the Quarkus `build` goal cacheable is isolated to the [QuarkusBuildCache](./src/main/java/com/gradle/QuarkusBuildCache.java) class.

Reviewer: sounds good

### Quarkus configuration dump
A key component of the caching mechanism is the Quarkus configuration dump file `.quarkus/quarkus-prod-config-dump`.
This file is generated by the Quarkus `build` goal when the Maven property `quarkus.config-tracking.enabled` is `true`.
It contains all the Quarkus properties used during the Quarkus `build` process.

> Some properties are discovered late in the `build` phase and can't be determined advance, thus the need for a first full execution to generate the file.

Reviewer 2: ```suggestion
Some properties are discovered late in the `build` phase and can't be determined in advance, thus the need for a first full execution to generate the file.
```

Author: Thanks, I fixed it

The presence of the file is required to mark the Quarkus `build` goal as cacheable. 

The `track-config-changes` goal creates a file `target/quarkus-prod-config-check` containing all the properties from the `.quarkus/quarkus-prod-config-dump` with their actual value.
If property values are identical in the two files, it means that the Quarkus configuration was not changed since the last Quarkus `build`, therefore the Quarkus `build` goal can be marked cacheable.

When the Quarkus `build` goal is marked cacheable, the regular caching process using [inputs](#goal-inputs) and [outputs](#goal-outputs) kicks in as described [here](https://docs.gradle.com/enterprise/maven-extension/#using_the_build_cache).

### Illustrated sequence of operations 
Let's illustrate the extension behavior with the following sequence of builds:

**During the first build:**
- `track-config-changes` does nothing as `.quarkus/quarkus-prod-config-dump` is absent
- Quarkus configuration from current and previous build differ

  => The `build` goal is not cacheable
- `build` executes and creates `.quarkus/quarkus-prod-config-dump`

> ![Run1](./doc/run-1.png)

Reviewer: nice to add these diagrams!
